src/operator/tensor/indexing_op.h

```diff
@@ -323,58 +323,89 @@ inline bool TakeOpShape(const nnvm::NodeAttrs& attrs,
 }
 
 inline bool TakeOpType(const nnvm::NodeAttrs& attrs,
-                       std::vector<int> *in_type,
-                       std::vector<int> *out_type) {
-  // using single dtype ("float32") for safety reason
-  CHECK_GE(in_type->size(), 2U);
-  int dtype = (*in_type)[1];
-  CHECK_NE(dtype, -1) << "idx must have specified type";
-  for (index_t i = 0; i < in_type->size(); ++i) {
-    if ((*in_type)[i] == -1) {
-      (*in_type)[i] = dtype;
-    } else {
-      CHECK_EQ((*in_type)[i], dtype) << "This layer requires uniform type. "
-                                     << "Expected " << dtype << " v.s. given "
-                                     << (*in_type)[i];
-    }
+                       std::vector<int> *in_attrs,
+                       std::vector<int> *out_attrs) {
+  CHECK_EQ(in_attrs->size(), 2U);
+  CHECK_EQ(out_attrs->size(), 1U);
+
+  TYPE_ASSIGN_CHECK(*out_attrs, 0, (*in_attrs)[0]);
+  TYPE_ASSIGN_CHECK(*in_attrs, 0, (*out_attrs)[0]);
+  if (-1 == (*in_attrs)[1]) {
+    TYPE_ASSIGN_CHECK(*in_attrs, 1, (*in_attrs)[0]);
   }
-  out_type->clear();
-  out_type->push_back(dtype);
-  return true;
+  return (*in_attrs)[0] != -1;
 }
 
+/*! \brief name the struct Take instead of take
+ * to avoid conflict with the take function in mshadow
+ */
+struct Take {
+  // assume that idx have been flattened to a 1-D tensor (N,)
+  // assume that out_data and in_data have been flattened to 2-D tensors, (N, M) and (K, M)
+  // M is the number of columns of in_data and out_data
+  // K is the number of rows of in_data
+  // i is the index of out_data
+  template<typename DType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, const DType* in_data,
+                                  const IType* idx, const int M, const int K) {
+    int j = static_cast<int>(idx[i/M]);
+    if (j <= 0) j = 0;
+    else if (j >= K) j = K - 1;
+    out_data[i] = in_data[j * M + i % M];
+  }
+};
+
+template<int req>
+struct TakeBackward {
+  // assume that idx have been flattened to a 1-D tensor (K,)
+  // assume that out_grad and in_grad have been flattened to 2-D tensors, (K, M) and (L, M)
+  // M is the number of columns of in_grad and out_grad
+  // K is the number of elements in idx
+  // L is the number rows of in_grad
+  // i is the index of in_grad
+  template<typename DType, typename IType>
+  MSHADOW_XINLINE static void Map(int i, DType* in_grad, const DType* out_grad,
```

Comment:
This is the loop of in_grad instead of out_grad. Race condition only arises when looping through out_grad. In addition, the original implementation uses the mshadow take(), which requires same DType of index and data array.